### DISCUSSION TITLE
Are Non Standard S3 capabilities ok for PRs?

### DISCUSSION THREAD

**Discussion by ibsusu** [seaweedfs/seaweedfs#2158] [Ideas]
I'm about to make an extension to the S3 protocol for 
`Content-SHA512`
`Content-SHA256`
`Content-Hash-Filename`

 `Content-Hash-Filename` just sets the requirement to match the non-padded base64 hash of the file to the file name of the file.  `Content-SHA512` and `Content-SHA256` are just alternatives to `Content-MD5`.  The reason I want this is to have a more confident way to presign urls.  I need it for my own uses so if you don't want it that's fine.

I'll be starting this file https://github.com/chrislusf/seaweedfs/blob/1cac2f2278b2344ac41ac882c7b21593f212d3bd/weed/storage/needle/needle_parse_upload.go#L92 and filling out the necessary bits from there.

If you have any input it would be appreciated.

**Comment by chrislusf** [comment]
S3 is provided to be compatible with AWS S3. So customization is unlikely mergeable.

`needle_parse_upload.go` is already at the blob storage layer. The presign URLs are done in S3 layer.

Existing code already has `X-Amz-Content-Sha256`
https://github.com/chrislusf/seaweedfs/blob/master/weed/s3api/auth_signature_v4.go#L61

**Reply by ibsusu** [reply]
I can use `X-Amz-Content-Sha256` and make my own `X-Amz-Content-Sha512` but that checksum isn't used anywhere else to validate the uploaded file.  md5 is though.  Am I thinking about this correctly?  Create the presigned url in auth_signature_v4.go with the `X-Amz-Content-xxx` and check it in `needle_parse_upload.go::ParseUpload`?  Or should I attempt to use something similar to `Content-MD5` in order to do an upload verification?

**Reply by ibsusu** [reply]
I think I understand, the `Content-MD5` is part of a standard whereas the `X-Amz-Content-Sha256` is not.  OK.  I think I know which way to go now.

**Reply by ibsusu** [reply]
Last thought but I find it a bit weird that something similar to https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-digest-headers

hasn't been implemented yet.  I guess because it's not ready?  Anyways, thanks for your input.

**Reply by lewisl9029** [reply]
Apparently S3 recently introduced algorithms other than md5 for integrity checking: https://aws.amazon.com/blogs/aws/new-additional-checksum-algorithms-for-amazon-s3/

Sha256 is one of them, and I'm similarly interested in support for sha256 checksums in Seaweed. 

It's one of the supported algorithms for https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity, so I could use the same checksum end-to-end all the way to browsers, which would be pretty awesome!
